lib/systems/default.nix

```diff
@@ -1,9 +1,87 @@
+{ lib
+}:
+let
+ inherit (lib.attrsets) mapAttrs;
+
+ selectEmulator = platform: pkgs:
+   let
+     qemu-user = pkgs.qemu.override {
+       smartcardSupport = false;
+       spiceSupport = false;
+       openGLSupport = false;
+       virglSupport = false;
+       vncSupport = false;
+       gtkSupport = false;
+       sdlSupport = false;
+       pulseSupport = false;
+       pipewireSupport = false;
+       smbdSupport = false;
+       seccompSupport = false;
+       enableDocs = false;
+       hostCpuTargets = [ "${platform.qemuArch}-linux-user" ];
+     };
+     wine = (pkgs.winePackagesFor "wine${toString platform.parsed.cpu.bits}").minimal;
+   in
+     if lib.systems.canExecute pkgs.stdenv.hostPlatform platform
+     then "${pkgs.runtimeShell} -c '\"$@\"' --"
+     else if platform.isWindows
+     then "${wine}/bin/wine${lib.optionalString (platform.parsed.cpu.bits == 64) "64"}"
+     else if platform.isLinux && pkgs.stdenv.hostPlatform.isLinux && platform.qemuArch != null
+     then "${qemu-user}/bin/qemu-${platform.qemuArch}"
+     else if platform.isWasi
+     then "${pkgs.wasmtime}/bin/wasmtime"
+     else if platform.isMmix
+     then "${pkgs.mmixware}/bin/mmix"
+     else null;
 
-rec {
   doubles = import ./doubles.nix { inherit lib; };
+
   parse = import ./parse.nix { inherit lib; };
+
+  # uncomment after 23.11 branch-off
+
+  /*
+  # These definitions have to be floated out into a `let … in` binding to ensure
+  # that they are the same exact value struct in the evaluating Nix
+  # implementation. Due to a quirk in how equality is implemented, values including
+  # functions are considered equal if they have the same memory location as long as
+  # they are inside a container value and not compared directly. Thus any attribute
+  # set will compare equal to itself even if it contains functions.
+  # By floating the functions out of the attribute set construction we'll ensure
+  # that the functions have the same memory location regardless of what call to
+  # lib.systems.elaborate produced them. This ensures that platform sets will be
+  # equal even if they have been constructed by different calls to lib.systems.elaborate
+  # —as long as their other contents are the same.
+  #
+  # For details see:
+  # - https://github.com/NixOS/nix/issues/3371
+  # - https://code.tvl.fyi/about/tvix/docs/value-pointer-equality.md
+  canExecutePlaceholder        = _: throw "2023-06-17: `platform.canExecute` has been removed in favor of `lib.systems.canExecute platform`";
+  emulatorAvailablePlaceholder = _: throw "2023-06-17: `platform.emulatorAvailable` has been removed in favor of `lib.systems.emulatorAvailable platform`";
+  emulatorPlaceholder          = _: throw "2023-06-17: `platform.emulator` has been removed in favor of `lib.systems.emulator platform`";
+  */
+
+  isCompatiblePlaceholder      = _: throw "2022-05-23: isCompatible has been removed in favor of canExecute, refer to the 22.11 changelog for details";
```

Comment:
Please change this to a warning.
By throwing, you force third party package expressions to insert compatibility logic in order to support the current releases.
This migration is not urgent, so we should not force third party packagers to write this kind of logic.

Reply:
@roberth I think you are missing some context here.

As you can see in the diff, `isCompatible` is already a hard error, from May.  There's no point in this PR bringing it _back_.

Reply:
I tried to review too quickly, and I jumped to a wrong conclusion here based on the title and earlier discussion, but not enough reading.